### PR TITLE
Use Scratch builds in Dockerfiles

### PR DESCRIPTION
**Pull request by BisonAl**
Closes https://bisontrails.atlassian.net/browse/PRO-102.

Changes introduced:
- Few small readme edits
- Move `build_and_bash` script up directly under `docker/`, tweak to be able to build any image with subdir argument
- Rewrite `bookkeeping` and `sync` node Dockerfiles, pulling all artifacts collected on `ubuntu:18.04` builder images into a final `scratch` image

Our image size for the `bookkeeping` node went from **325MB** to **18.9MB**, really stoked about that 🎉
 
Example `sync` build:
```shell
al@Alexs-MacBook-Pro docker (scratch_builds) $ bash build_and_bash.sh sync_node/
Sending build context to Docker daemon  3.584kB
Step 1/6 : FROM ubuntu:18.04 as builder
 ---> 2ca708c1c9cc
Step 2/6 : RUN apt-get update     && apt-get upgrade -y     && apt-get install -yq apt-utils     && apt-get install -yq build-essential     && apt-get install -yq ca-certificates      && apt-get install -yq curl
 ---> Using cache
 ---> ed722c0d3a3c
Step 3/6 : RUN curl https://dev.ont.io/ontology_install | sh
 ---> Using cache
 ---> f35c097205f4
Step 4/6 : FROM scratch
 ---> 
Step 5/6 : COPY --from=builder /ontio /ontio
 ---> e2b8f1724772
Step 6/6 : ENTRYPOINT ["/ontio/ontology"]
 ---> Running in 9557a1661592
Removing intermediate container 9557a1661592
 ---> f86c5d96df6a
Successfully built f86c5d96df6a
Successfully tagged ontology:latest
2019/10/17 15:51:27.574633 [INFO ] GID 1, ontology version v1.7.2-0-g2377759
2019/10/17 15:51:27.574921 [INFO ] GID 1, Config init success
2019/10/17 15:51:27.590192 [INFO ] GID 1, deploy contract address:0239dcf9b4a46f15c5f23f20d52fac916a0bac0d
2019/10/17 15:51:27.590511 [INFO ] GID 1, deploy contract address:08b6dcfed2aace9190a44ae34a320e42c04b46ac
2019/10/17 15:51:27.590614 [INFO ] GID 1, deploy contract address:7a2f84e3b94d20da1a8592116c0103c28c5e457e
2019/10/17 15:51:27.590782 [INFO ] GID 1, deploy contract address:6815cbe7b4dbad4d2d09ae035141b5254a002f79
2019/10/17 15:51:27.590830 [INFO ] GID 1, deploy contract address:24a15c6aed092dfaa711c4974caf1e9d307bf4b5
2019/10/17 15:51:27.590868 [INFO ] GID 1, deploy contract address:4d6934f0a524a084bb20cff4cdbea236760bb4a3
2019/10/17 15:51:27.592855 [INFO ] GID 1, GenesisBlock init success. GenesisBlock hash:1b8fa7f242d0eeb4395f89cbb59e4c29634047e33245c4914306e78a88e14ce5
...
```

Example `bookkeeping` build (wallet error expected):
```shell
al@Alexs-MacBook-Pro docker (scratch_builds) $ bash build_and_bash.sh bookkeeping_node/
Sending build context to Docker daemon  3.584kB
Step 1/7 : FROM ubuntu:18.04 as builder
 ---> 2ca708c1c9cc
Step 2/7 : RUN apt-get update 	&& apt-get upgrade -y 	&& apt-get install -yq apt-utils 	&& apt-get install -yq build-essential 	&& apt-get install -yq ca-certificates  	&& apt-get install -yq curl
 ---> Using cache
 ---> 9cf272b67b8c
Step 3/7 : RUN curl https://dev.ont.io/ontology_install | sh
 ---> Using cache
 ---> e54570c1b7f2
Step 4/7 : RUN echo "{}" >> /ontio/wallet.dat
 ---> Using cache
 ---> 23fcc3dd8ba2
Step 5/7 : FROM scratch
 ---> 
Step 6/7 : COPY --from=builder /ontio /ontio
 ---> Using cache
 ---> b1bffec1aae0
Step 7/7 : ENTRYPOINT ["/ontio/ontology", "--enable-consensus", "--disable-rpc", "--disable-event-log"]
 ---> Using cache
 ---> 686df493d3eb
Successfully built 686df493d3eb
Successfully tagged ontology:latest
2019/10/17 15:55:18.802857 [INFO ] GID 1, ontology version v1.7.2-0-g2377759
2019/10/17 15:55:18.803175 [INFO ] GID 1, Config init success
2019/10/17 15:55:18.803452 [ERROR] GID 1, initWallet error: Cannot find wallet file: ./wallet.dat. Please create a wallet first
```

### CODE REVIEW

File: docker/bookkeeping_node/Dockerfile [hunk 1]
```diff
@@ -15,6 +15,13 @@ RUN apt-get update \
...
+## secret manager system, just allows us
+## to fail at a reasonable point for now.
+RUN echo "{}" >> /ontio/wallet.dat
 
-ENTRYPOINT ["sh", "entrypoint.sh"]
\ No newline at end of file
+# Copy ubuntu artifacts to new scratch container
+FROM scratch
+COPY --from=builder /ontio /ontio
+ENTRYPOINT ["/ontio/ontology", "--enable-consensus", "--disable-rpc", "--disable-event-log"]
```
Comment (andrewbaine) [inline docker/bookkeeping_node/Dockerfile]: should those flags be moved to:
```
CMD ["--enable-consensus", ...]

File: docker/build_and_bash.sh [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+# of a subdirectory, and runs the 
+# entrypoint command
+
+if [ ! $1 ]; then
+    echo "Please choose a subdirectory as arg1"
+    exit 1
+fi
+
+
+docker build -t ontology ./$1
```
Comment (andrewbaine) [inline docker/build_and_bash.sh]: it seems dangerous to tag a bunch of different nodes as ontology. Is there a difference in the binary that each of the different node types runs? If not, then  I'd propose get rid of all but one directory, and use ENTRYPOINT and CMD and wrapper bash scripts to invoke the docker image with flags that are appropriate for each mode. Basically i'm asking, do we need separate directories witin `docker`?

Comment (BisonAl) [reply docker/build_and_bash.sh]: That's a really good idea, I'll do that now!

File: docker/run_bookkeeper.sh [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# Run ontology image, with
+# processing CMD args and therefore
+# running as a bookkeeper node
+docker build -t ontology .
+docker run -it ontology
```
Comment (andrewbaine) [inline docker/run_bookkeeper.sh]: get rid of '-it' and exec in if you need shell access?

File: docker/run_sync.sh [hunk 4]
```diff
@@ -0,0 +1,5 @@
+# Run ontology image, without
+# processing CMD args and therefore
+# running as a sync node
+docker build -t ontology .
+docker run -it ontology /bin/bash
```
Comment (andrewbaine) [inline docker/run_sync.sh]: does this run the `ontology` binary or `/bin/bash` ? I think you should get rid of the '-it' and exec, but that's just my opinion

File: docker/Dockerfile [hunk 5]
```diff
@@ -24,4 +24,5 @@ RUN echo "{}" >> /ontio/wallet.dat
 # Copy ubuntu artifacts to new scratch container
 FROM scratch
 COPY --from=builder /ontio /ontio
-ENTRYPOINT ["/ontio/ontology", "--enable-consensus", "--disable-rpc", "--disable-event-log"]
+ENTRYPOINT ["/ontio/ontology"]
+CMD ["--enable-consensus", "--disable-rpc", "--disable-event-log"]
```
Comment (andrewbaine) [inline docker/Dockerfile]: now that you've fixed everything, maybe `CMD []` should be empty and the flags should be added to the modes that need them; cause I don't know how to replace CMD with empty array at runtime...

Comment (BisonAl) [reply docker/Dockerfile]: Yeah I like that more, will do 👍

**Comment by BisonAl** [comment]
Addressed your last round of comments with https://github.com/bisontrails/network-ontology/pull/5/commits/f74f9f8e67e40348331264ec2b8b3be15a46cbac @andrewbaine, taking your advice and shipping now too 👍

**Pull request merged**